ckan/views/__init__.py

```diff
@@ -43,27 +50,142 @@ def set_cors_headers_for_response(response: Response) -> Response:
+            if 'etag' not in response.headers:
+                # s3 etag uses md5, so using it here also
+
+                content_type = response.mimetype or ''
+                allowed_types = {'text/plain', 'text/css',
+                                 'text/html', 'application/json'}
+                etag_super_strong = u'__etag_super_strong__' in request.environ
+
+                try:
+                    # anon/public get ignored csrf
+                    if (etag_super_strong
+                            or content_type not in allowed_types
+                            or current_user.is_authenticated or g.user):
+                        data_to_hash = response.get_data()
+                    else:
+                        # Regex for both _csrf_token content and csp nonce and don't
+                        # care if there is new lines spacing etc attributes which are
+                        # dynamic on pages
+                        # May need to add more when we come across them.
+
+                        field_name = re.escape(config.get("WTF_CSRF_FIELD_NAME", "_csrf_token"))  # noqa: E501
+                        # csrf meta only, if to include value,
+                        #   update (?:content) to (?:content|value)
+                        pattern = fr'(?i)((?:_csrf_token|{field_name})[^>]*?\b(?:content)=|\bnonce=)["\'][^"\']+(["\'])'  # noqa: E501
+
+                        # Replace values with etag_removed
+                        response_data = re.sub(pattern,
+                                               lambda m: m.group(1) + '="etag_removed"',
+                                               response.get_data(as_text=True))
+                        data_to_hash = response_data.encode()
+
+                    etag = hashlib.md5(data_to_hash).hexdigest()
+                    response.set_etag(etag)
+                except (AttributeError, IndexError, TypeError,
+                        UnicodeEncodeError, ValueError, re.error) as e:
+                    logging.info("Failed to compute and set ETag: %s", e)
+            else:
+                etag = response.headers.get('etag')
+
+            etag_not_conditional = u'__etag_not_conditional__' not in request.environ
+            # Allow legacy behaviour if config is set
+            if enable_etags_not_modified and etag_not_conditional:
+                # Use built-in function now that we have an eTag
+                response.make_conditional(request.environ)
+
+    return response
+
+
 def set_cache_control_headers_for_response(response: Response) -> Response:
+    """ This uses the presents of request environ's: '__no_cache__',
+    '__no_private_cache__', '__limit_cache_by_cookie__' as well
+    as config variables to control cache response headers"""
+
+    is_webasset = u'__webasset__' in request.environ
+    if is_webasset:
+        # Don't alter web assets as flask_app has handled cache control for us
+        return response
```

Comment:
instead of a webasset-specific check can we instead check that the response already has some cache-control headers set? That would be more generic and allow custom views to be more intelligent.

Reply:
if it's in the `g` object then it can be set or unset anywhere, including e.g. from some template snippet that was included in a theme.

Maybe that's useful? e.g. a theme might add a feature to a page that is dynamic and shouldn't be cached.

If it can be set anywhere should we prevent loosening the cache setting once it's set for a particular request in case it's modified in more than one place? @smotornyuk what do you think?